### PR TITLE
ROE-1938 - TL - Gov BO et al - Focus required for empty date error

### PR DESCRIPTION
**Pull request by mattch1**
### JIRA link
https://companieshouse.atlassian.net/browse/ROE-1938

### Change description

Mainly refactoring: 

Moved empty date check to the day field validation to provide the click and focus required.
Replaced existing check in the test for the whole date with a simple true or false check and appended it to the other checks.

Also noticed here that a condition was being repeated for start dates for the work done on ROE-1987: https://github.com/companieshouse/overseas-entities-web/pull/694/files

 if (checkMoreThanOneDateFieldIsNotMissing(dayStr, monthStr, yearStr) && isYearEitherMissingOrCorrectLength(yearStr)) {
    if (isYearEitherMissingOrCorrectLength(yearStr)) {

so removed the unnecessary extra check.

As the changes affected the optional date and date of birth so have had to update these as well to be consistent with the start date.

We now have separate day field checks for date of birth, start date and optional/identity date. Moreover, the month and year checks need another parameter as dob has a different message.

### Work checklist

- [x] Tests added where applicable
- [x] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Review by SVarallo — APPROVED** [review]
Did some testing on the branch and seemingly fixes https://companieshouse.atlassian.net/browse/ROE-1936 as well. Looks all good to me but would recommend getting a second opinion

File: src/validation/custom.validation.ts [hunk 1]
```diff
@@ -29,13 +29,20 @@ export const checkInvalidCharactersIfRadioButtonSelected = (selected: boolean, e
   return true;
 };
 
-export const checkDateIsNotCompletelyEmpty = (errMsg: string, day: string = "", month: string = "", year: string = "") => {
+export const checkDateIsNotCompletelyEmptyWithErrorThrown = (errMsg: string, day: string = "", month: string = "", year: string = "") => {
```
Comment (mwestacott) [inline src/validation/custom.validation.ts:32]: I think now having two functions with the same logic in apart from either throwing an error or returning false is not great. Better to have just the one method (keep the one below) which returns true or false and in those three (I think) cases where it's currently called and needs to throw the error, react to the 'false' value returned and do that there.

File: src/validation/custom.validation.ts [hunk 2]
```diff
@@ -151,25 +152,55 @@ const checkStartDateFields = (dayStr: string = "", monthStr: string = "", yearSt
...
+export const checkDateFieldDayOfBirth = (dayStr: string = "", monthStr: string = "", yearStr: string = "") => {
+  if (isYearEitherMissingOrCorrectLength(yearStr)) {
+    if (dayStr === "" && monthStr !== "" && yearStr !== "") {
+      throw new Error(ErrorMessages.DAY_OF_BIRTH);
+    } else if (dayStr === "" && monthStr === "" && yearStr !== "") {
+      throw new Error(ErrorMessages.DAY_AND_MONTH_OF_BIRTH);
+    } else if (dayStr === "" && monthStr !== "" && yearStr === "") {
+      throw new Error(ErrorMessages.DAY_AND_YEAR_OF_BIRTH);
+    } else {
+      checkDateIsNotCompletelyEmptyWithErrorThrown(ErrorMessages.ENTER_DATE_OF_BIRTH, dayStr, monthStr, yearStr);
```
Comment (mwestacott) [inline src/validation/custom.validation.ts:179]: See comment above. So here you'd have to check for 'false' being returned and if so throw the error...

File: src/validation/custom.validation.ts [hunk 3]
```diff
@@ -151,25 +152,55 @@ const checkStartDateFields = (dayStr: string = "", monthStr: string = "", yearSt
...
+    } else if (dayStr === "" && monthStr === "" && yearStr !== "") {
+      throw new Error(ErrorMessages.DAY_AND_MONTH);
+    } else if (dayStr === "" && monthStr !== "" && yearStr === "") {
+      throw new Error(ErrorMessages.DAY_AND_YEAR);
+    }
+  }
+  return true;
+};
+
+export const checkDateFieldMonth = (monthMissingMessage: string, monthYearMsg: string, dayStr: string = "", monthStr: string = "", yearStr: string = "") => {
```
Comment (mwestacott) [inline src/validation/custom.validation.ts:198]: Would make the parameter naming consistent here - so `monthMissingMessage` and `monthYearMissingMessage`.

File: src/validation/custom.validation.ts (hunk 1 above)
Comment (mattch1) [reply src/validation/custom.validation.ts]: Removed second function

File: src/validation/custom.validation.ts (hunk 2 above)
Comment (mattch1) [reply src/validation/custom.validation.ts]: Removed second function

File: src/validation/custom.validation.ts (hunk 3 above)
Comment (mattch1) [reply src/validation/custom.validation.ts]: Renamed

**Review by mwestacott — APPROVED** [review]

**Pull request merged**